### PR TITLE
enhancement(external docs): Provide templating for different condition syntax options

### PR DESCRIPTION
**Pull request by lucperkins**
This PR replaces the changes proposed in #9091. It both incorporates those changes and also modifies the CUE schemas and HTML templating to present more granular information about conditions. The templating is a bit quick and dirty but I think it's a clear improvement.

The affected pages:

* https://deploy-preview-9307--vector-project.netlify.app/docs/reference/configuration/transforms/filter
* https://deploy-preview-9307--vector-project.netlify.app/docs/reference/configuration/transforms/route
* https://deploy-preview-9307--vector-project.netlify.app/docs/reference/configuration/transforms/sample

### CODE REVIEW

File: website/cue/reference.cue [hunk 1]
```diff
@@ -464,7 +467,21 @@ _values: {
...
 	}
 
-	syntax: "file_system_path" | "field_path" | "literal" | "template" | "regex" | "remap_boolean_expression" | "remap_program" | "strftime"
+	syntax: #StringSyntax
+}
+
+#TypeCondition: {
+	required: bool
+
+	#ConditionSyntax: "datadog_search" | "vrl_boolean_expression"
```
Comment (jszwedko) [inline website/cue/reference.cue:476]:
```suggestion
	#ConditionSyntax: "datadog_search" | "vrl"
```

The `type` is actually just `vrl` in the config loading code:

https://github.com/vectordotdev/vector/blob/e10294c7b41b7328f708c03a37740c9e617270d8/src/conditions/vrl.rs#L23

File: website/cue/reference.cue [hunk 2]
```diff
@@ -436,6 +437,8 @@ _values: {
 }
 
 #TypeString: {
+	#StringSyntax: *"literal" | "file_system_path" | "field_path" | "template" | "regex" | "remap_program" | "strftime"
```
Comment (jszwedko) [inline website/cue/reference.cue:440]:
```suggestion
	#StringSyntax: *"literal" | "file_system_path" | "field_path" | "template" | "regex" | "vrl_program" | "strftime"
```

File: website/cue/reference/components/transforms/filter.cue [hunk 3]
```diff
@@ -42,11 +43,18 @@ components: transforms: filter: {
 				"""
 			required: true
 			warnings: []
-			type: string: {
+			type: condition: {
+				syntax: ["datadog_search", "vrl_boolean_expression"]
```
Comment (jszwedko) [inline website/cue/reference/components/transforms/filter.cue:47]:
```suggestion
				type: ["datadog_search", "vrl_boolean_expression"]
```

I believe this is actually called `type` in the config:

https://github.com/vectordotdev/vector/blob/e10294c7b41b7328f708c03a37740c9e617270d8/src/conditions/mod.rs#L31

File: website/cue/reference/components/transforms/filter.cue [hunk 4]
```diff
@@ -67,7 +75,7 @@ components: transforms: filter: {
 		{
 			title: "Drop debug logs"
 			configuration: {
-				condition: #".level == "debug""#
+				condition: #".level != "debug""#
```
Comment (jszwedko) [inline website/cue/reference/components/transforms/filter.cue:78]: 👍

**Review by jszwedko — COMMENTED** [review]
Nice thanks for tackling this @lucperkins . It is a definite improvement.

I left a few comments. Also, I think it might be a little confusing that the `.type` and `.source` field are not literally called out in the config doc like other nested options are. What would you think of doing something like we do with the `encoding` option on sinks where each field is listed independently:

Screen Shot 2021-09-23 at 1 31 45 PM

?

I am wondering if we should mention the short-hand VRL filter syntax too where you can just do `condition: ".foo == 3"`. We could maybe just put this in the description of the `vrl` condition type?

File: website/cue/reference/components/transforms/filter.cue (hunk 4 above)
Comment (lucperkins) [reply website/cue/reference/components/transforms/filter.cue:78]: Hehe, yeah, this was carried over from @leebenson's PR

**Comment by lucperkins** [comment]
@jszwedko I guess I'm missing something here. For the `filter` transform, can you specify either just a string or an object or do you have to specify an object with `source` and `type` given that there are two options?

**Comment by jszwedko** [comment]
> @jszwedko I guess I'm missing something here. For the `filter` transform, can you specify either just a string or an object or do you have to specify an object with `source` and `type` given that there are different syntax options?

Yeah, it is a little confusing. Basically you can specify VRL conditions in two ways:

```toml
condition = '.level != "debug'`
```

Or:

```toml
condition.type = "vrl"
condition.source = '.level != "debug"'
```

`datadog_search` conditions can only be specified like:

```toml
condition.type = "datadog_search"
condition.source = '*something*'
```

The VRL condition short-hand make more sense before we introduced the new type. I would say you could just document the expanded form:

```toml
condition.type = "vrl"
condition.source = '.level != "debug"'
```

And then mention in the description that there is a short-hand form of just specifying `condition` as a string rather than an object.

**Comment by lucperkins** [comment]
Yeah, so, I've spent hours on this without a really solid way forward. I'm gonna come up with something kind of ham-fisted for it but in the future I think we should seriously just allowing some of these configs to be plain Markdown text. It's hard to justify the continued time expenditure.

**Comment by jszwedko** [comment]
@lucperkins what would you think of reviving this? It came up again today in the context of @jdrouet's work on adding conditions to pipelines.

**Comment by lucperkins** [comment]
@jszwedko I can take a look at this, yeah. But I've been really struggling finding a reasonable way to present this and I think this we should be more circumspect about this kind of "flexible" configuration in the future.

**Comment by lucperkins** [comment]
I don't like the approach I initially ran with here and it's been long enough that rebasing with `master` is pretty thorny. I'm going to close and start from scratch.

**Pull request closed without merging**